### DISCUSSION TITLE
Watching dependencies in generated project

### DISCUSSION THREAD

**Discussion by tpluscode** [copier-org/copier#593] [Q&A] [answered]
I started building templates for node projects but similar issues could occur with any target.

Possibly, there will be a template for the file which defined dependencies, such as `package.json.jinja`. This prevents the usual tooling which integrates with GitHub from functioning. Dependabot, Greenkeeper or Renovate will not find the dependencies to upgrade

Does anyone have a strategy for automating how and when versions of dependencies are updated in the templates?

**Comment by pawamoy** [comment]
I've been using a somewhat dumb script that generates a project, goes into it, run some tool that updates the dependencies where they are specified (in my case `pyproject.toml`, in yours `package.json`), and outputs the new dependencies so I can paste them back in the template myself. This can probably be fully automated though. And I don't have the script anymore but you get the idea :slightly_smiling_face: 

The other solution is simply to use [an empty suffix for templates](https://copier.readthedocs.io/en/latest/configuring/#templates_suffix) in your projects. But Dependabot and friends will still likely choke on Jinja syntax if you use it in the file anyway.

**Comment by cjolowicz** [comment]
I wrote [retrocookie](https://retrocookie.readthedocs.io) to solve this problem in the Cookiecutter world. You can use it to copy commits or PRs from a generated project to its template. I'm using it to import Dependabot PRs for the Hypermodern Python template, from a special [instance repository](https://github.com/cjolowicz/cookiecutter-hypermodern-python-instance). It currently doesn't support copier syntax, but it shouldn't be hard to generalize if anybody wants to work on this.

**Reply by tpluscode** [reply]
Aha, that would be neat. 

In fact, does it even have to concern itself with the syntax? I was not even thinking in jinja terms. In what I had in mind it would have simply taken to cherry-pick the changes to `package.json` back to the original file. Either named thusly, or a `package.json.jinja`

**Reply by cjolowicz** [reply]
You could set up a GitHub workflow that

1. exports the patch using `git format-patch`
2. rewrites the path for the template repository
3. imports the patch using `git am` into the template repository, 
4. pushes to the template repository

It's quite a bit of work figuring out the details, but it can definitely be done.

**Comment by yajo** [comment]
I can think on various solutions. I will post them in different comments so they can get voted separately.

For example, you can use [tasks](https://copier.readthedocs.io/en/stable/configuring/#tasks) to generate the file dynamically after copying.

**Reply by tpluscode** [reply]
Ideally, I would have the file as `package.json` so that dependant can update the dependency versions.

If I kept Jinja tags, would it be possible to run Jinja for that one file from a task? (I do not want to set an empty template extension)

**Reply by yajo** [reply]
I doubt that dependant would work fine with a json file full of Jinja expressions.

**Reply by tpluscode** [reply]
Might also remove expressions and patch the json in tasks entirely

**Comment by yajo** [comment]
Just using Jinja, with a smart combination of include, import, macro, set and the from_json filter, you could recover the package.json updated by dependabot. Then use that into your template.

**Comment by tpluscode** [comment, marked as answer]
Ok, I did eventually as mentioned in [comment](https://github.com/copier-org/copier/discussions/593#discussioncomment-2337277)

1. `package.json` are not jinja templates ([example](https://github.com/hypermedia-app/creta-template-app/blob/master/src/package.json))
2. I update them from command line in [tasks](https://github.com/hypermedia-app/creta-template-app/blob/master/copier.yaml#L20-L30) using [`npx json`](http://npm.im/json)

This makes dependabot's business as usual
